apps/igloo-kit-cli/src/framework/controller.rs

```diff
@@ -0,0 +1,245 @@
+use crate::framework::languages::CodeGenerator;
+use crate::infrastructure::olap::clickhouse::ClickhouseTable;
+use crate::infrastructure::stream;
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::framework::languages::SupportedLanguages;
+
+use crate::framework;
+
+use log::debug;
+use tokio::sync::Mutex;
+
+use crate::framework::typescript::get_typescript_models_dir;
+
+use crate::framework::typescript::SendFunction;
+
+use crate::framework::sdks::TypescriptObjects;
+
+use crate::project::Project;
+
+use std::path::PathBuf;
+
+use crate::infrastructure::olap;
+
+use std::io::ErrorKind;
+
+use crate::infrastructure::olap::clickhouse::ClickhouseView;
+
+use std::io::Error;
+
+use crate::infrastructure::olap::clickhouse::ConfiguredDBClient;
+
+use super::schema::parse_schema_file;
+use super::schema::MatViewOps;
+use super::schema::Table;
+use super::schema::TableOps;
+use super::typescript::TypescriptInterface;
+
+pub struct FrameworkObject {
+    pub table: ClickhouseTable,
+    pub topic: String,
+    pub ts_interface: TypescriptInterface,
+}
+
+pub fn framework_object_mapper(t: Table) -> FrameworkObject {
+    let clickhouse_table = olap::clickhouse::mapper::std_table_to_clickhouse_table(t.clone());
+    FrameworkObject {
+        table: clickhouse_table.clone(),
+        topic: t.name.clone(),
+        ts_interface: framework::typescript::mapper::std_table_to_typescript_interface(t),
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RouteMeta {
+    pub original_file_path: PathBuf,
+    pub table_name: String,
+    pub view_name: Option<String>,
+}
+
+pub fn get_framework_objects(route: &PathBuf) -> Result<Vec<FrameworkObject>, Error> {
+    let framework_objects =
+        parse_schema_file::<FrameworkObject>(route.clone(), framework_object_mapper).map_err(
+            |e| {
+                Error::new(
+                    ErrorKind::Other,
+                    format!("Failed to parse schema file. Error {}", e),
+                )
+            },
+        )?;
+    Ok(framework_objects)
+}
+
+pub(crate) async fn create_or_replace_view(
+    fo: &FrameworkObject,
+    view_name: String,
+    configured_client: &ConfiguredDBClient,
+) -> Result<(), Error> {
+    let view = ClickhouseView::new(fo.table.db_name.clone(), view_name, fo.table.clone());
+    let create_view_query = view.create_materialized_view_query().map_err(|e| {
+        Error::new(
+            ErrorKind::Other,
+            format!("Failed to get clickhouse query: {:?}", e),
+        )
+    })?;
+
+    // Clickhouse doesn't support dropping a view if it doesn't exist so we need to drop it first in case the schema has changed
+    let drop_view_query = view.drop_materialized_view_query().map_err(|e| {
```

Comment:
I don't get the algorithm probably something I am missing.

- Creating table
- Dropping the table 
- Creating the same table again

Why the 2 first steps? Or is I that if the table already exists it not going to do anything?

Reply:
Not sure if I'm seeing what you're seeing. The algorithm is:

1. Drop table if it exists
2. Create table if it does not exist

This should create the table only if it does not already exist and replace it if it does. 

Note that in the line you're highlighting, we're only creating the query, which we then run. Maybe you're seeing the query creation and running as the separate steps?

Reply:
Can you point me to the `if exists`  logic? I guess that's what I am missing.

Reply:
Line 103 in /infrastructure/olap/clickhouse/queries.rs

Reply:
Ahh - I just realized one statement is to create the query while the other executes it.  🤦🏻 .

Thank you for bearing with me